server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java

```diff
@@ -370,9 +374,13 @@ public IndexFieldData.Builder getFieldDataBuilder(String name) {
 
             private static void validateParsed(float value) {
                 if (Float.isFinite(HalfFloatPoint.sortableShortToHalfFloat(HalfFloatPoint.halfFloatToSortableShort(value))) == false) {
-                    throw new IllegalArgumentException("[half_float] supports only finite values, but got [" + value + "]");
+                    throwOnInfiniteHalfFloat(value);
```

Comment:
Is this important to the optimization? I think I'd prefer to leave this as is unless there is a measurable effect?

Reply:
Hard to say. It certainly gives the compiler a better shot at generating good code than not doing this.

This is what current master looks like when profiling:

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/6490959/165071097-32c1d0ec-341d-4812-b419-09a86436c072.png">

Obviously, fixing the intermediary list is the biggest saving we can make but it's fairly visible that just the pure byte parsing side of things isn't free either CPU wise.
You can pretty easily make this a visible effect in a JMH benchmark. Whether or not this will translate into a real-world effect will very much depend on the use case (mapping) and environment.
It's rather hard to make this kind of thing show in a rally run because of how noisy they are.

That said, my thinking as far whether it's worth it optimising this kind of code path to this degree would be:

We spend roughly half the CPU time on the write thread during indexing on parsing documents in benchmarks like the Rally logging track. There's not that all that many obvious optimisations in this code (the list thing here being a notable exception) but it's slower in terms of bytes/s than it could theoretically be. In the past optimisations like this around x-content have been quite successful and Jackson itself goes to great lengths to do the same optimisations on their end.
Each individual change like this will not necessarily show up much outside of JMH but the aggregate effects of these changes will eventually by allowing the compiler to optimise with greater scope after inlining more.
-> while admittedly there's always a little voodoo to this kind of optimisation and a single digit % change in the performance of this code is hard to measure in Rally, I'd still go for it simply because this is among the hottest code paths we have and any bit of saving here will translate into real world cost savings for users now and might reveal optimisations to us (and/or the compiler :)) down the line.

Reply:
I think I'd like to leave out these minor optimizations from this PR and then have a separate discussion on those. I think we need a firmer stance on the necessity of this and if we end up needing this, we should have some verification that we are not inadvertently violating this in our tests or build. We should probably discuss that in a meeting in the coming weeks. But I think it is worthwhile getting the main optimization (the intermediary list optimization) in, so would like to decouple the two.

Reply:
Sounds good, reverted the JIT optimizations for now then :)